### ISSUE TITLE
ToggleTermSendVisualSelection doesn't work as expected

### ISSUE THREAD

**Issue by etjupm** [akinsho/toggleterm.nvim#235]
The ToggleTermSendVisualSelection command sends the block defined by the start and end of the visual selection rather than the visual selection itself. For example, with the lines:
```
echo hello
echo world
```
attempting to select and send the whole line `echo hello` and then `echo w` will end up sending `echo h\necho w`. Sending to the terminal will fail altogether in visual-block mode if the starting column is to the left of the ending column.

The problem seems to be in the M.send_lines_to_terminal function in lua/toggleterm.lua, specifically the following block:
```lua
  if selection_type == "visual_lines" or selection_type == "visual_selection" then
    local res = line_selection("visual")
    b_line, b_col = unpack(res.start_pos)
    lines = res.selected_lines

    if selection_type == "visual_selection" then
      -- Visual selection is more accurate, as we get the sub-string of every line based on the visual selection
      local _, e_col = unpack(res.end_pos)
      for i, v in ipairs(lines) do
        lines[i] = v:sub(b_col, e_col)
      end
    end
  elseif selection_type == "single_line" then
    b_line, b_col = unpack(api.nvim_win_get_cursor(0))
    table.insert(lines, fn.getline(b_line))
  end
```

In my local version of the plugin, I have replaced that block with the following, which uses the vim.fn.visualmode() function to get the previous visual mode used and then execute the correct procedure for each type of visual mode:
```lua
  if selection_type == "single_line" then
    b_line, b_col = unpack(api.nvim_win_get_cursor(0))
    table.insert(lines, fn.getline(b_line))

  elseif selection_type == "visual_lines" then
    local res = line_selection("visual")
    b_line, b_col = unpack(res.start_pos)
    lines = res.selected_lines

  elseif selection_type == "visual_selection" then
    local res = line_selection("visual")
    b_line, b_col = unpack(res.start_pos)
    local vis_mode = fn.visualmode()

    if vis_mode == "V" then
      -- line-visual
      lines = res.selected_lines

    elseif vis_mode == "v" then
      -- regular-visual
      local start_line, start_col = unpack(res.start_pos)
      local end_line, end_col = unpack(res.end_pos)
      lines = api.nvim_buf_get_text(
        0, start_line - 1, start_col - 1, end_line, end_col, {}
      )

    elseif vis_mode == "\x16" then
      -- block-visual
      local _, start_col = unpack(res.start_pos)
      local _, end_col = unpack(res.end_pos)
      if start_col > end_col then
        -- exchange values for proper string substitution if needed
        start_col, end_col = end_col, start_col
      end
      for i, v in ipairs(res.selected_lines) do
        lines[i] = v:sub(start_col, end_col)
      end
    end
  end
```

This seems to mostly work, although the style may not be to your liking. The one case where it won't work as expected is if selection is set to "exclusive", since I can't find a way to test for that in lua. In that case, the regular-visual selection will include the next character after the selection, as if selection was "inclusive".

**Comment by akinsho** [comment]
Hi @etjupm,

So this functionality was kindly submitted by another user and I suggested that rather than put it in the wiki it might deserve being part of the plug-in. I think it is by no means perfect and probably will never be since there are so many edge cases. I personally don't add to or tweak this functionality since it's kind of out of scope for this plugin but if you or another user wants to improve the logic for the better then I really don't mind. I don't really use or actively maintain it in either case

**Comment by akinsho** [comment]
closed by #239
